pkg/controller/controller_binding.go

```diff
@@ -577,55 +580,93 @@ func (c *controller) reconcileServiceBinding(binding *v1beta1.ServiceBinding) er
+		toUpdate, err := makeServiceBindingClone(binding)
 		if err != nil {
-			s := fmt.Sprintf(`Error deleting secret: %s`, err)
-			glog.Warning(pcb.Message(s))
-			c.recorder.Eventf(binding, corev1.EventTypeWarning, errorEjectingBindReason, "%v %v", errorEjectingBindMessage, s)
+			return err
+		}
+
+		instance, err := c.instanceLister.ServiceInstances(binding.Namespace).Get(binding.Spec.ServiceInstanceRef.Name)
+		if err != nil {
+			s := fmt.Sprintf(
+				`References a non-existent %s "%s/%s"`,
+				pretty.ServiceInstance, binding.Namespace, binding.Spec.ServiceInstanceRef.Name,
+			)
+			glog.Warningf(pcb.Messagef("%s (%s)", s, err))
+			c.recorder.Event(binding, corev1.EventTypeWarning, errorNonexistentServiceInstanceReason, s)
 			setServiceBindingCondition(
 				toUpdate,
 				v1beta1.ServiceBindingConditionReady,
-				v1beta1.ConditionUnknown,
-				errorEjectingBindReason,
-				errorEjectingBindMessage+s,
+				v1beta1.ConditionFalse,
+				errorNonexistentServiceInstanceReason,
+				"The binding references an ServiceInstance that does not exist. "+s,
 			)
 			if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
 				return err
 			}
 			return err
 		}
 
-		unbindRequest := &osb.UnbindRequest{
-			BindingID:  binding.Spec.ExternalID,
-			InstanceID: instance.Spec.ExternalID,
-			ServiceID:  serviceClass.Spec.ExternalID,
-			PlanID:     servicePlan.Spec.ExternalID,
+		if instance.Status.AsyncOpInProgress {
+			s := fmt.Sprintf(
+				`trying to bind to %s "%s/%s" that has ongoing asynchronous operation`,
+				pretty.ServiceInstance, binding.Namespace, binding.Spec.ServiceInstanceRef.Name,
+			)
+			glog.Info(pcb.Message(s))
+			c.recorder.Event(binding, corev1.EventTypeWarning, errorWithOngoingAsyncOperation, s)
+			setServiceBindingCondition(
+				toUpdate,
+				v1beta1.ServiceBindingConditionReady,
+				v1beta1.ConditionFalse,
+				errorWithOngoingAsyncOperation,
+				errorWithOngoingAsyncOperationMessage,
+			)
+			if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
+				return err
+			}
+			return fmt.Errorf("Ongoing Asynchronous operation")
 		}
 
-		if utilfeature.DefaultFeatureGate.Enabled(scfeatures.OriginatingIdentity) {
-			originatingIdentity, err := buildOriginatingIdentity(binding.Spec.UserInfo)
-			if err != nil {
-				s := fmt.Sprintf(`Error building originating identity headers while unbinding: %v`, err)
-				glog.Warning(pcb.Message(s))
-				c.recorder.Event(binding, corev1.EventTypeWarning, errorWithOriginatingIdentity, s)
-				setServiceBindingCondition(
-					toUpdate,
-					v1beta1.ServiceBindingConditionReady,
-					v1beta1.ConditionFalse,
-					errorWithOriginatingIdentity,
-					s,
-				)
-				if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
-					return err
-				}
+		deprovisionRequired := true
+		if instance.Spec.ClusterServiceClassRef == nil || instance.Spec.ClusterServicePlanRef == nil {
```

Comment:
This is going to leave a binding in the broker in the case where the binding is deleted after the plan was changed on the instance and before the controller successfully resolves the plan reference.

Reply:
how likely is that? and can we use this PR as a pivot point to add similar DeprovisionStatus -> UnbindStatus as a followup?

Reply:
I think it is very likely. It would occur if the user tries to change to a non-existent plan.

Reply:
My concern is that we this PR is added some marginal protection for one scenario in a way that is very specific to that scenario. We know that it is not a finished solution and not even the husk of the solution that we think we want. So I see one of two things happening. Either (1) we do the complete solution soon, in which case this PR is wasted because it does not move us meaningfully in that direction. Or (2) we do not do the complete solution soon, forget why this special case is here, and carry code debt for it moving forward.